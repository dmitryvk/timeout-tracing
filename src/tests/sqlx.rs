use std::time::Duration;

use serial_test::serial;
use sqlx::{Connection, PgConnection};
use tracing::instrument;

use crate::{
    TimeoutElapsed,
    tests::{insta_trace_filters, run_with_tracing},
};

#[tokio::test]
#[ignore]
#[serial]
async fn with_sqlx() {
    let result = run_with_tracing(Duration::from_millis(1000), do_sqlx()).await;

    assert!(matches!(result, Err(TimeoutElapsed { .. })));
    let traces = &result.as_ref().err().unwrap().active_traces;
    for (trace_idx, trace) in traces.iter().enumerate() {
        assert!(
            trace.span_trace.to_string().contains("exec_query"),
            "trace #{trace_idx}"
        );
        assert!(
            trace.stack_trace.to_string().contains("exec_query"),
            "trace #{trace_idx}"
        );
    }
    assert!(
        traces
            .iter()
            .any(|e| e.stack_trace.to_string().contains("exec_query")
                && e.span_trace.to_string().contains("pg_sleep(2)")
                && e.span_trace.to_string().contains("exec_query")),
        "no trace contains exec_query"
    );
    insta::with_settings!({
        filters => insta_trace_filters()
    }, {
        insta::assert_debug_snapshot!(result);
    });
}

#[instrument]
async fn do_sqlx() {
    let mut conn = sqlx::PgConnection::connect("postgresql://pg:123@localhost:15432/pg")
        .await
        .unwrap();
    exec_query(&mut conn, "select pg_sleep(0.001)").await;
    exec_query(&mut conn, "select pg_sleep(2)").await;
}

#[instrument(skip(conn))]
async fn exec_query(conn: &mut PgConnection, sql: &str) {
    let _results: Vec<()> = sqlx::query_as(sql).fetch_all(conn).await.unwrap();
}
