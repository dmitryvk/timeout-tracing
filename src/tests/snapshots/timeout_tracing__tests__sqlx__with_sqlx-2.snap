---
source: src/tests/sqlx.rs
expression: err
---
timeout elapsed at:
trace 0:
span trace:
   0: timeout_tracing::tests::sqlx::exec_query
           with sql="select pg_sleep(2)"[crates]/serial_test-[ver]/src/serial_code_lock.rs:[NNN]:[NNN]
  75: timeout_tracing::tests::sqlx::with_sqlx
             at ./src/tests/sqlx.rs:[NNN]:[NNN]
  76: timeout_tracing::tests::sqlx::with_sqlx::{{closure}}
             at ./src/tests/sqlx.rs:[NNN]:[NNN]
  77: core::ops::function::FnOnce::call_once
  78: core::ops::function::FnOnce::call_once
  79: test::__rust_begin_short_backtrace
  80: test::run_test_in_process::{{closure}}
  81: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  82: std::panicking::try::do_call
  83: std::panicking::try
  84: std::panic::catch_unwind
  85: test::run_test_in_process
  86: test::run_test::{{closure}}
  87: test::run_test::{{closure}}
  88: std::sys::backtrace::__rust_begin_short_backtrace
  89: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
  90: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
  91: std::panicking::try::do_call
  92: std::panicking::try
  93: std::panic::catch_unwind
  94: std::thread::Builder::spawn_unchecked_::{{closure}}
  95: core::ops::function::FnOnce::call_once{{vtable.shim}}
  96: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
  97: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
  98: std::sys::pal::unix::thread::Thread::new::thread_start
  99: <unknown>
 100: <unknown>
